src/overcooked_ai_py/mdp/layout_generator.py

```diff
@@ -12,74 +12,156 @@
 DISH_DISPENSER = 'D'
 SERVING_LOC = 'S'
 
-CODE_TO_TYPE = { 0: EMPTY, 1: COUNTER, 2: ONION_DISPENSER, 3: TOMATO_DISPENSER, 4: POT, 5: DISH_DISPENSER, 6: SERVING_LOC}
-TYPE_TO_CODE = { v:k for k, v in CODE_TO_TYPE.items() }
+CODE_TO_TYPE = {0: EMPTY, 1: COUNTER, 2: ONION_DISPENSER, 3: TOMATO_DISPENSER, 4: POT, 5: DISH_DISPENSER,
+                6: SERVING_LOC}
+TYPE_TO_CODE = {v: k for k, v in CODE_TO_TYPE.items()}
+
+
+
+def mdp_fn_random_choice(mdp_fn_choices):
+    assert type(mdp_fn_choices) is list and len(mdp_fn_choices) > 0
+    return random.choice(mdp_fn_choices)
+
+
+"""
+size_bounds: (min_layout_size, max_layout_size)
+prop_empty: (min, max) proportion of empty space in generated layout
+prop_feats: (min, max) proportion of counters with features on them
+"""
+
+DEFAULT_MDP_GEN_PARAMS = {
+    "inner_shape": (5, 4),
+    "prop_empty": 0.95,
+    "prop_feats": 0.1,
+    "start_all_orders" : [
+        { "ingredients" : ["onion", "onion", "onion"]}
+    ],
+    "recipe_values" : [20],
+    "recipe_times" : [20],
+    "display": False
+}
+
+
+def DEFAILT_PARAMS_SCHEDULE_FN(outside_information):
+    mdp_default_gen_params = {
+        "inner_shape": (5, 4),
+        "prop_empty": 0.95,
+        "prop_feats": 0.1,
+        "start_all_orders" : [
+        { "ingredients" : ["onion", "onion", "onion"]}
+        ],
+        "recipe_values" : [20],
+        "recipe_times" : [20],
+        "display": False
+    }
+    return mdp_default_gen_params
+
+
+class MDPParamsGenerator(object):
+
+    def __init__(self, params_schedule_fn=None, mdp_params_always=None):
+        """
+        outer_shape: the outer shape of the mdp to be generated
+        schedule_fn: the function to produce a set of mdp_params for a specific layout
+        mdp_params_always: the mdp_params that the generator should always been using (no scheduling)
+        """
+        assert params_schedule_fn != None or mdp_params_always != None, \
+            "one of params_schedule_fn and mdp_params_always must be not Null"
+        self.mdp_params_always = mdp_params_always
```

Comment:
mdp_params_always can be considered a special case of params_schedule_fn where the schedule always returns the same params (similar to the mdp_generator_fn). Up to you whether you want to make this change – I don't have strong opinions about how the layout generator should be organized

Reply:
I left this in to be backwards compatible (because some part of the old code calls `mdp_gen_fn_from_dict` using the mdp_params dictionary), especially if we want to use it to generate mdp for evaluators

Reply:
We don't actually care about backwards compatibility for layoutgenerator code as it was never really used anywhere

Reply:
I am sorry, I think I misread your initial comment, and made an ambiguous response because of my understanding. I will addressed this in the newest commit.

Reply:
Ah no worries! :)